contracts/zaps/ZapCurveTokenToPrincipalToken.sol

```diff
@@ -0,0 +1,458 @@
+                    );
+
+                    // Due to rounding issues of some tokens, we find the
+                    // relevant token balance of this contract
+                    _ctx[i] = IERC20(_zap.roots[i]).balanceOf(address(this));
+                }
+            }
+        }
+
+        // When we do not have anything to swap for, we short-circuit
+        if (!zapHasAmounts && !ctxHasAmounts) {
+            return 0;
+        }
+
+        // It is necessary to add liquidity to the respective curve pool like
+        // this due to the non-standard interface of the function in the curve
+        // contracts. Not only is there two variants of fixed length array
+        // amount inputs but it is often inconsistent whether return values
+        // exist or not
+        // Also, the context amounts container, although fixed-length of 2 or 3,
+        // because it is a low-level function call, solidity will not complain
+        // and the first 2 indexes are only considered in the case of it being
+        // length 2
+        address(_zap.curvePool).functionCallWithValue(
+            abi.encodeWithSelector(_zap.funcSig, _ctx, 0),
+            msg.value
+        );
+
+        return _zap.lpToken.balanceOf(address(this));
+    }
+
+    /// @dev zapIn enables the user swap multiple dependent curve tokens into
+    /// their respective lp tokens and then immediately swaps them for
+    /// principal tokens, greatly simplifying the purchasing of principal
+    /// tokens
+    /// @param _info See ZapInInfo struct
+    /// @param _zap See ZapCurveLpIn struct - This is the "main" or parent zap
+    /// which produces the lp token necessary to swap for the principal token
+    /// @param _childZap See ZapCurveLpIn - This is used only in cases where
+    /// the "main" or "parent" zap itself is composed of another curve lp token
+    /// which can be accessed more readily via another swap via curve
+    function zapIn(
+        ZapInInfo memory _info,
+        ZapCurveLpIn memory _zap,
+        ZapCurveLpIn memory _childZap,
+        PermitData[] memory _permitData
+    )
+        external
+        payable
+        notFrozen
+        preApproval(_permitData)
+        returns (uint256 ptAmount)
+    {
+        // Instantiation of the context amount container which is used to track
+        // amounts to be swapped in the final curve zap.
+        uint256[3] memory ctx;
+
+        // Only execute the childZap if it is necessary
+        if (_info.needsChildZap) {
+            uint256 _amount = _zapCurveLpIn(
+                _childZap,
+                // The context array is unnecessary for the childZap and so we
+                // can just put a dud array in place of it
+                [uint256(0), uint256(0), uint256(0)]
+            );
+            // When a childZap happens, we add the amount of lpTokens gathered
+            // from it to the relevant root index of the "main" zap
+            ctx[_childZap.parentIdx] += _amount;
```

Comment:
We could use the `coins` method on the curve contract maybe? It takes an integer value returning the address of the erc20 at that index. We could validate if the returned address from that call matches the childZap LP token? But we also would run into the issue of different interfaces for the `coins` method so that would require passing func sig for that from off-chain.